simulation_test.go

```diff
@@ -0,0 +1,169 @@
+	if len(supporting_hashes) != len(execProof.Nodes()) {
+		return false
+	}
+	for idx, supporting_hash := range supporting_hashes {
+		modelMinNS := int(supporting_hash.Get("minNS").Int())
+		execMinNs := bytesToInt(execProof.Nodes()[idx][:nidSize])
+		if modelMinNS != execMinNs {
+			return false
+		}
+
+		modelMaxNS := int(supporting_hash.Get("maxNS").Int())
+		execMaxNs := bytesToInt(execProof.Nodes()[idx][nidSize:(2 * nidSize)])
+		if modelMaxNS != execMaxNs {
+			return false
+		}
+	}
+
+	return true
+
+}
+
+func intToBytes(n int, bytesSize int) []byte {
+	b := make([]byte, bytesSize)
+	for i := 0; i < bytesSize; i++ {
+		b[i] = byte(n >> (8 * (bytesSize - i - 1)))
+	}
+	return b
+}
+
+func bytesToInt(bytes []byte) int {
+	var result int
+	for i, b := range bytes {
+		shift := uint((len(bytes) - 1 - i) * 8)
+		result |= int(b) << shift
+	}
+	return result
+}
+
+func TestFromITF(t *testing.T) {
+
+	var tree *nmt.NamespacedMerkleTree
+
+	itfFileName := "formal_spec/ITF_files/runTest.itf.json"
+	// itfFileName := "formal_spec/ITF_files/panicIssue.itf.json"
+	data, err := ioutil.ReadFile(itfFileName)
+	if err != nil {
+		t.Errorf("Error opening file: %v", err)
+	}
+	states := gjson.GetBytes(data, "states").Array()
+
+	// iterate over all states of the test run
+	for stateIdx, state := range states {
+		t.Logf("State %v\n", stateIdx)
+		// because the test run consists of multiple steps to generate the tree and the proof, test
+		// only the states marked "final"
+		if state.Get("state_v").String() == "final" {
+			modelNamespace := int(state.Get("namespace_v").Int())
+			modelLeaves := getLeavesFromState(state, nidSize)
+			modelProof := state.Get("proof_v")
+			corrupted := state.Get("corrupted").Bool()
+			t.Logf("Obtained state data:\n\tState namespace: %v\n\tleaves: %v\n\tcorrupted: %v\n",
+				modelNamespace, modelLeaves, corrupted)
+
+			tree = nmt.New(sha256.New(), nmt.NamespaceIDSize(nidSize))
+			for _, leaf := range modelLeaves {
+				err := tree.Push(leaf)
+				if err != nil {
+					t.Errorf("Error on push: %v", err)
+				}
+			}
+
+			namespaceBytes := intToBytes(modelNamespace, nidSize)
+
+			proof, err := tree.ProveNamespace(namespaceBytes)
+			if err != nil {
+				t.Errorf("Error on prove: %v", err)
+			}
+			var new_hashes [][]byte
+			t.Logf("Proof: %v\n", proof)
+
+			var corruption_type string
+			if !corrupted {
+				proofMatching := checkProof(modelProof, proof, nidSize)
+				if !proofMatching {
+					t.Errorf("Proofs do not match. Expected: %v, got: %v", state, proof)
+				}
+			} else {
+				// CORRUPTED: now we are modifying the proof to be corrupted here as well
+				corruption_type = state.Get("corruption_type").String()
+				t.Log("Corruption type: ", corruption_type)
+				if corruption_type == "supporting_hashes" {
+					for idx := 0; idx < len(state.Get("corruption_diff.changed_indices").Array()); idx++ {
+						new_hashes = append(new_hashes, proof.Nodes()[idx])
+					}
+				} else {
```

Comment:
I am curious why the other types of corruption are not being examined to modify only the part of the generated `proof` relevant to the specific corruption type. Currently, all fields are being read from the `modelProof`. However, depending on the type of corruption, only a specific field of the `proof` should be altered, not all of them.

Reply:
I believe you are right, we could use only the corrupted part from the `modelProof` and the rest from `proof`. However, reading everything from `modelProof` will not cause harm because `modelProof` should match `proof` in all but corrupted fields.

Reply:
> will not cause harm because modelProof should match proof in all but corrupted fields.

I see, It seems like this is an implicit assumption which is not clear immediately from the go code (e.g., there is no prior field by field comparison of the `modelProof` and the `proof`). To ensure better clarity, it may be beneficial to add a comment in the code or revise the code itself to only modify the relevant parts of the proofs  that are affected by the corruption case.